File: composeApp/src/androidMain/kotlin/com/ivangarzab/kluvs/ui/components/AttendanceControl.kt
package com.ivangarzab.kluvs.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.dp
import com.ivangarzab.kluvs.R
import com.ivangarzab.kluvs.model.AttendanceResponse
import com.ivangarzab.kluvs.model.AttendanceRoster
import com.ivangarzab.kluvs.model.AttendanceStatus
import com.ivangarzab.kluvs.theme.KluvsTheme
import com.ivangarzab.kluvs.theme.statusDanger
import com.ivangarzab.kluvs.theme.statusDangerSubtle
import com.ivangarzab.kluvs.theme.statusSuccess
import com.ivangarzab.kluvs.theme.statusSuccessSubtle

private val SEGMENTS = listOf(AttendanceStatus.YES, AttendanceStatus.MAYBE, AttendanceStatus.NO)

/**
 * RSVP control for a single discussion — mirrors web's `AttendanceControl`.
 *
 * A 3-segment icon pill (yes, maybe, no); tapping the already-selected segment
 * clears the RSVP. Renders nothing until [roster] is loaded.
 */
@Composable
fun AttendanceControl(
    roster: AttendanceRoster?,
    disabled: Boolean,
    onSetAttendance: (AttendanceStatus) -> Unit,
    modifier: Modifier = Modifier
) {
    if (roster == null) return

    val counts = SEGMENTS.associateWith { status -> roster.responses.count { it.status == status } }

    Column(modifier = modifier) {
        Row(
            modifier = Modifier
                .clip(RoundedCornerShape(50))
                .border(1.dp, MaterialTheme.colorScheme.outlineVariant, RoundedCornerShape(50))
                .alpha(if (disabled) 0.7f else 1f),
            verticalAlignment = Alignment.CenterVertically
        ) {
            SEGMENTS.forEachIndexed { index, status ->
                AttendanceSegment(
                    status = status,
                    isSelected = roster.myStatus == status,
                    disabled = disabled,
                    isFirst = index == 0,
                    onClick = { onSetAttendance(status) }
                )
            }
        }
        Text(
            text = "${counts[AttendanceStatus.YES]} yes · ${counts[AttendanceStatus.NO]} no · ${counts[AttendanceStatus.MAYBE]} maybe",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 4.dp)
        )
    }
}

@Composable
private fun AttendanceSegment(
    status: AttendanceStatus,
    isSelected: Boolean,
    disabled: Boolean,
    isFirst: Boolean,
    onClick: () -> Unit
) {
    val (background, tint) = when {
        isSelected && status == AttendanceStatus.YES -> statusSuccessSubtle to statusSuccess
        isSelected && status == AttendanceStatus.NO -> statusDangerSubtle to statusDanger
        isSelected -> MaterialTheme.colorScheme.surfaceVariant to MaterialTheme.colorScheme.onSurface
        else -> MaterialTheme.colorScheme.surface to MaterialTheme.colorScheme.onSurfaceVariant
    }

    Row(
        modifier = Modifier
            .size(28.dp)
            .then(
                if (!isFirst) {
                    Modifier.border(
                        width = 1.dp,
                        color = MaterialTheme.colorScheme.outlineVariant
                    )
                } else Modifier
            )
            .background(background)
            .clickable(enabled = !disabled, onClick = onClick),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically
    ) {
        when (status) {
            AttendanceStatus.YES -> Icon(
                imageVector = Icons.Default.Check,
                contentDescription = "RSVP yes",
                tint = tint,
                modifier = Modifier.size(13.dp)
            )

            AttendanceStatus.MAYBE -> Icon(
                painter = painterResource(R.drawable.ic_help),
                contentDescription = "RSVP maybe",
                tint = tint,
                modifier = Modifier.size(13.dp)
            )

            AttendanceStatus.NO -> Icon(
                imageVector = Icons.Default.Close,
                contentDescription = "RSVP no",
                tint = tint,
                modifier = Modifier.size(13.dp)
            )
        }
    }
}

@PreviewLightDark
@Composable
private fun Preview_AttendanceControl() = KluvsTheme {
    AttendanceControl(
        roster = AttendanceRoster(
            responses = listOf(
                AttendanceResponse(memberId = "0", name = "Ivan", status = AttendanceStatus.YES),
                AttendanceResponse(memberId = "1", name = "Sam", status = AttendanceStatus.MAYBE)
            ),
            myStatus = AttendanceStatus.YES,
            totalMembers = 6
        ),
        disabled = false,
        onSetAttendance = {}
    )
}

File: iosApp/iosApp/Components/AttendanceControl.swift
import SwiftUI
import Shared

/// RSVP control for a single discussion — mirrors web's `AttendanceControl` /
/// Android's `AttendanceControl`. A 3-segment icon pill (yes, maybe, no);
/// tapping the already-selected segment clears the RSVP (handled by the
/// ViewModel, which treats a re-select as a clear). Renders nothing until
/// `roster` is loaded.
struct AttendanceControl: View {
    let roster: Shared.AttendanceRoster?
    let disabled: Bool
    let onSetAttendance: (Shared.AttendanceStatus) -> Void

    private static let segments: [Shared.AttendanceStatus] = [.yes, .maybe, .no]

    var body: some View {
        if let roster {
            VStack(alignment: .leading, spacing: 4) {
                HStack(spacing: 0) {
                    ForEach(Array(Self.segments.enumerated()), id: \.offset) { index, status in
                        AttendanceSegment(
                            status: status,
                            isSelected: roster.myStatus == status,
                            disabled: disabled,
                            isFirst: index == 0,
                            onTap: { onSetAttendance(status) }
                        )
                    }
                }
                .overlay(
                    RoundedRectangle(cornerRadius: 14)
                        .strokeBorder(Color(uiColor: .separator), lineWidth: 1)
                )
                .clipShape(RoundedRectangle(cornerRadius: 14))
                .opacity(disabled ? 0.7 : 1.0)

                Text(countsLabel(for: roster))
                    .font(.caption2)
                    .foregroundColor(.secondary)
                    .monospacedDigit()
            }
        }
    }

    private func countsLabel(for roster: Shared.AttendanceRoster) -> String {
        let yes = roster.responses.filter { $0.status == .yes }.count
        let no = roster.responses.filter { $0.status == .no }.count
        let maybe = roster.responses.filter { $0.status == .maybe }.count
        return "\(yes) yes · \(no) no · \(maybe) maybe"
    }
}

private struct AttendanceSegment: View {
    let status: Shared.AttendanceStatus
    let isSelected: Bool
    let disabled: Bool
    let isFirst: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            icon
                .font(.system(size: 13, weight: .semibold))
                .foregroundColor(tint)
                .frame(width: 28, height: 28)
                .background(background)
                .overlay(alignment: .leading) {
                    if !isFirst {
                        Rectangle()
                            .fill(Color(uiColor: .separator))
                            .frame(width: 1)
                    }
                }
        }
        .disabled(disabled)
    }

    private var icon: Image {
        switch status {
        case .yes: Image(systemName: "checkmark")
        case .maybe: Image(systemName: "questionmark")
        case .no: Image(systemName: "xmark")
        default: Image(systemName: "questionmark")
        }
    }

    private var background: Color {
        guard isSelected else { return .clear }
        switch status {
        case .yes: return .statusSuccessSubtle
        case .no: return .statusDangerSubtle
        default: return Color(uiColor: .secondarySystemFill)
        }
    }

    private var tint: Color {
        guard isSelected else { return .secondary }
        switch status {
        case .yes: return .statusSuccess
        case .no: return .statusDanger
        default: return .primary
        }
    }
}

#Preview {
    AttendanceControl(
        roster: Shared.AttendanceRoster(
            responses: [
                Shared.AttendanceResponse(memberId: "0", name: "Ivan", status: .yes),
                Shared.AttendanceResponse(memberId: "1", name: "Sam", status: .maybe)
            ],
            myStatus: .yes,
            totalMembers: 6
        ),
        disabled: false,
        onSetAttendance: { _ in }
    )
    .padding()
}
